File: Domains/InfiniteMario/15-rl-competition-2009/system/codecs/Java/src/org/rlcommunity/rlglue/codec/tests/Test_Speed_Environment.java
/*
 * Copyright 2008 Brian Tanner
 * http://rl-glue-ext.ext.googlecode.com/
 * brian@tannerpages.com
 * http://brian.tannerpages.com
 * 
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.rlcommunity.rlglue.codec.tests;

import org.rlcommunity.rlglue.codec.EnvironmentInterface;
import org.rlcommunity.rlglue.codec.util.EnvironmentLoader;
import org.rlcommunity.rlglue.codec.types.Action;
import org.rlcommunity.rlglue.codec.types.Observation;
import org.rlcommunity.rlglue.codec.types.Reward_observation_terminal;

/**
 *
 * @author Brian Tanner
 */
public class Test_Speed_Environment implements EnvironmentInterface {


   int whichEpisode=0;
   int stepCount=0;
   
   Observation o =new Observation();
    
    public Test_Speed_Environment() {
    }

    
    public String env_message(String inMessage) {
        return "";
    }

    public static void main(String[] args){
        EnvironmentLoader L=new EnvironmentLoader(new Test_Speed_Environment());
        L.run();
    }

    public String env_init() {
        return "";
    }
    public Observation env_start() {
        stepCount=0;
        whichEpisode++;
        TestUtility.clean_abstract_type(o);
        return o;   
    }

    public Reward_observation_terminal env_step(Action action) {
        stepCount++;
        Reward_observation_terminal ro=null;
        
        TestUtility.clean_abstract_type(o);
        
        //Short episode with big observations
        if(whichEpisode%2==0){
            TestUtility.set_k_ints_in_abstract_type(o, 50000);
            TestUtility.set_k_doubles_in_abstract_type(o, 50000);

            int terminal=0;
            if(stepCount==200)terminal=1;
                ro=new Reward_observation_terminal(1.0d, o, terminal);
        }
        //Longer episode with smaller obserations
        if(whichEpisode%2==1){
            TestUtility.set_k_ints_in_abstract_type(o, 5);
            TestUtility.set_k_doubles_in_abstract_type(o, 5);

            int terminal=0;
            if(stepCount==5000)terminal=1;
                ro=new Reward_observation_terminal(1.0d, o, terminal);
        }
        
                
        return ro;
    }

    public void env_cleanup() {
    }
}

File: Domains/InfiniteMario/15-rl-competition-2009/system/codecs/C/tests/test_speed_environment.c
/* 
* Copyright (C) 2007, Brian Tanner

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.

* 
*  $Revision: 365 $
*  $Date: 2008-11-04 01:19:46 -0500 (Tue, 04 Nov 2008) $
*  $Author: brian@tannerpages.com $
*  $HeadURL: http://rl-glue-ext.googlecode.com/svn/trunk/projects/codecs/C/tests/test_speed_environment.c $
* 
*/
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif


#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <rlglue/Environment_common.h>
#include <rlglue/utils/C/RLStruct_util.h>
#include "useful_functions.h"


static observation_t *o=0;
static reward_observation_terminal_t ro={0};
static int stepCount=0;
static int episodeCount=0;

const char* env_init()
{    
	o=allocateRLStructPointer(0,0,0);
	return "sample task spec";
}

const observation_t *env_start()
{
	episodeCount++;
	stepCount=0;
	clearRLStruct(o);
	__RL_CHECK_STRUCT(o)
	return o;
}

const reward_observation_terminal_t *env_step(const action_t *a)
{
	int terminal=0;
	stepCount++;
	clearRLStruct(o);
	    
        /*Short episode with big observations*/
        if(episodeCount%2==0){
			__RL_CHECK_STRUCT(o)
            set_k_ints_in_abstract_type(o, 50000);
			__RL_CHECK_STRUCT(o)
            set_k_doubles_in_abstract_type(o, 50000);
			__RL_CHECK_STRUCT(o)

            if(stepCount==200)terminal=1;
        }
        /*Longer episode with smaller observations*/
        if(episodeCount%2==1){
			__RL_CHECK_STRUCT(o)
            set_k_ints_in_abstract_type(o, 5);
		__RL_CHECK_STRUCT(o)
            set_k_doubles_in_abstract_type(o, 5);
		__RL_CHECK_STRUCT(o)

            if(stepCount==5000)terminal=1;
        }

	
	ro.observation=o;
	ro.reward=1.0;
	ro.terminal=terminal;
	__RL_CHECK_STRUCT(ro.observation)
    return &ro;
}

void env_cleanup()
{
  freeRLStructPointer(o);
}


const char* env_message(const char* inMessage) {
	return "";
}
